### PR TITLE
Fix calling convention propagation for function pointers.

### PR DESCRIPTION
**Pull request by emilio**
This sucks, but works.

The full solution is a refactoring that needs more thought than the time I'm
able to dedicate to bindgen right now, see the comment for details.

r? @fitzgen

### CODE REVIEW

**Comment by emilio** [comment]
I need to figure out a way for this to be green in CI given we're "cross-compiling" the test.

File: tests/headers/derive-fn-ptr.h [hunk 1]
```diff
@@ -0,0 +1,8 @@
+typedef void (*my_fun_t)(int, int, int, int,
+                         int, int, int, int,
+                         int, int, int, int,
+                         int, int, int, int);
```
Comment (fitzgen) [inline tests/headers/derive-fn-ptr.h]: Nice :-|

File: src/ir/function.rs [hunk 2]
```diff
@@ -349,3 +350,26 @@ impl Trace for FunctionSig {
...
 }
+
+// Function pointers follow special rules, see:
+//
+// https://github.com/servo/rust-bindgen/issues/547,
+// https://github.com/rust-lang/rust/issues/38848,
+// and https://github.com/rust-lang/rust/issues/40158
+//
+// Note that copy is always derived, so we don't need to implement it.
+impl CanDeriveDebug for FunctionSig {
```
Comment (fitzgen) [inline src/ir/function.rs]: At least extending these checks is obvious and straight forward.

**Comment by emilio** [comment]
> Since we generate different bindings with MSVC mangling anyways, it would make sense to have different expectations dirs for different targets. Then we could add appveyor CI.

Yeah, but then we need to add an easy way of updating test expectations across different targets. It seems just giving clang the right `-target` arg might just do it.

**Pull request merged**